### PR TITLE
feat(lib/es2015): Add typed overloads to `Reflect`

### PR DESCRIPTION
**Pull request by ExE-Boss**
This adds typed overloads to `Reflect`, similar to what `strictBindCallApply` does.

~~I’ve also corrected the type of `getPrototypeOf(…)` and `setPrototypeOf(…)`.~~ (extracted into <https://github.com/microsoft/TypeScript/pull/41987>)

## Depends on:
- [x] <https://github.com/microsoft/TypeScript/pull/38967> (merged)
- [x] <https://github.com/microsoft/TypeScript/pull/41987> (merged)

### CODE REVIEW

File: src/lib/es2015.reflect.d.ts [hunk 1]
```diff
@@ -1,15 +1,43 @@
...
-    function apply(target: Function, thisArgument: any, argumentsList: ArrayLike<any>): any;
-    function construct(target: Function, argumentsList: ArrayLike<any>, newTarget?: any): any;
+    /**
+     * Calls the function with the specified object as the this value and the elements of specified array as the arguments.
+     * @param target The function to call.
+     * @param thisArgument The object to be used as the this object.
+     * @param argumentsList An array of argument values to be passed to the function.
+     */
+    function apply<T, A extends any[], R>(target: (this: T, ...args: A) => R, thisArgument: T, argumentsList: A): R;
+    function apply(target: (...args: any) => any, thisArgument: any, argumentsList: ArrayLike<any>): any;
```
Comment (Jessidhia) [inline src/lib/es2015.reflect.d.ts]: This second overload will just be picked by the compiler whenever anything fails to be accepted by the first overload, so this doesn't really add strictness. Better autocomplete, I guess.

File: src/lib/es2015.reflect.d.ts [hunk 2]
```diff
@@ -1,15 +1,43 @@
...
+
+    /**
+     * Constructs the target with the elements of specified array as the arguments and the specified constructor as the `new.target` value.
+     * @param target The function to call.
+     * @param argumentsList An array of argument values to be passed to the function.
+     * @param newTarget The constructor to be used as the `new.target` object.
+     */
+    function construct<A extends any[], R>(target: new (...args: A) => R, argumentsList: A, newTarget?: new (...args: any) => any): R;
+    function construct(target: new (...args: any) => any, argumentsList: ArrayLike<any>, newTarget?: new (...args: any) => any): any;
```
Comment (Jessidhia) [inline src/lib/es2015.reflect.d.ts]: Same here, the second overload makes the first one useless :cry:

File: src/lib/es2015.reflect.d.ts [hunk 3]
```diff
@@ -1,15 +1,43 @@
...
+    /**
+     * Returns the prototype of an object.
+     * @param o The object that references the prototype.
+     */
+    function getPrototypeOf(target: object): object | null;
     function has(target: object, propertyKey: PropertyKey): boolean;
     function isExtensible(target: object): boolean;
     function ownKeys(target: object): PropertyKey[];
     function preventExtensions(target: object): boolean;
     function set(target: object, propertyKey: PropertyKey, value: any, receiver?: any): boolean;
```
Comment (Jessidhia) [inline src/lib/es2015.reflect.d.ts]: `set` / `has` / `get` could benefit from overloads like the ones made for `apply` and `construct`, though; `get` in particular.

File: src/lib/es2015.reflect.d.ts (hunk 1 above)
Comment (ExE-Boss) [reply src/lib/es2015.reflect.d.ts]: Unfortunately, this&nbsp;is&nbsp;needed to&nbsp;be&nbsp;able to&nbsp;pass the&nbsp;`arguments`&nbsp;object.

File: src/lib/es2015.iterable.d.ts [hunk 4]
```diff
@@ -214,10 +214,6 @@ interface PromiseConstructor {
     race<T>(values: Iterable<T | PromiseLike<T>>): Promise<T>;
 }
 
-declare namespace Reflect {
-    function enumerate(target: object): IterableIterator<any>;
-}
```
Comment (ExE-Boss) [inline src/lib/es2015.iterable.d.ts]: The&nbsp;`enumerate(…)`&nbsp;method was&nbsp;removed in&nbsp;**ES2016** from&nbsp;both&nbsp;`Reflect` and&nbsp;`ProxyHandler`s, due&nbsp;to&nbsp;implementers deciding&nbsp;against&nbsp;it.

Comment (ExE-Boss) [reply src/lib/es2015.iterable.d.ts:219]: Extracted to: <https://github.com/microsoft/TypeScript/pull/38967>

**Review by sandersn — APPROVED** [review]
This seems like a reasonable improvement that wouldn't break anything *obvious* because it leaves the originals behind as overloads. I did a quick sample of usage and `Reflect` is used sparingly across many prominent packages, so I think we'd notice problems after merging if there were going to be any.

However, I'd like to get the opinion of one other team member before merging, since lib changes nearly always have hidden gotches. @andrewbranch or @rbuckton can you give your opinion?

**Pull request merged**